caffe2/operators/expand_op.h

```diff
@@ -0,0 +1,107 @@
+#ifndef CAFFE2_OPERATORS_REDUCE_OPS_H_
+#define CAFFE2_OPERATORS_REDUCE_OPS_H_
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/core/types.h"
+#include "caffe2/utils/math.h"
+
+namespace caffe2 {
+
+template <typename InputTypes, class Context>
+class ExpandOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+
+  ExpandOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws) {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<InputTypes>::call(this, Input(0));
+  }
+ template <typename T>
+  bool DoRunWithType() {
+    const auto& X = Input(0);
+    const auto& Y_shape_tensor = Input(1);
+    std::vector<int64_t> shape_dims(Y_shape_tensor.size());
+    context_.template Copy<int64_t, Context, CPUContext>(
+        Y_shape_tensor.size(),
+        Y_shape_tensor.template data<int64_t>(),
+        shape_dims.data());
+    auto* Y = Output(0);
+
+	const int ndim = shape_dims.size();
+    const std::vector<int> X_dims(X.dims().cbegin(), X.dims().cend());
+    std::vector<int> Y_dims;
+    Y_dims.reserve(std::max(ndim, X.ndim()));
+    // ndim, X.ndim() might equal to 0
+    for (int i = ndim - 1, j = X.ndim() - 1; i >= 0 || j >= 0; i--, j--) {
```

Comment:
This block should be as simple as copying the Input(1) to Y_dims, the validation work should be done in math::Broadcast already.

Reply:
Maybe not? Or we can have another function to calculate shape. I think it is not a very good idea to let math::Broadcast handle too many things, but better just let it accept some standard data?